source/extensions/health_checkers/mysql/mysql.cc

```diff
@@ -0,0 +1,323 @@
+      user_(mysql_config.user()) {}
+
+MySQLHealthChecker::MySQLActiveHealthCheckSession::~MySQLActiveHealthCheckSession() {
+  if (client_) {
+    client_->close(Network::ConnectionCloseType::NoFlush);
+  }
+}
+
+template <size_t Bytes, typename Result = uint64_t>
+static inline Result read_little_endian_int(const uint8_t* data) {
+  static_assert(Bytes <= sizeof(Result), "possible integer overflow");
+
+  Result result = 0;
+
+  // we can't assume unaligned reads are permitted so we trust the standard library to efficiently
+  // perform the copy for small-sized buffers
+  std::copy(data, data + Bytes, reinterpret_cast<uint8_t*>(&result));
+
+  return from_little_endian(result);
+}
+
+template <size_t Bytes, typename T>
+static inline void write_little_endian_int(Buffer::Instance& out, T value) {
+  static_assert(Bytes <= sizeof(value), "possible integer underflow");
+
+  value = to_little_endian(value);
+  out.add(&value, Bytes);
+}
+
+constexpr const auto mysql_payload_length_offset = 0;
+constexpr const size_t mysql_payload_length_size = 3;
+
+constexpr const auto mysql_sequence_id_offset =
+    mysql_payload_length_offset + mysql_payload_length_size;
+constexpr const size_t mysql_sequence_id_size = 1;
+
+constexpr const size_t mysql_packet_header_size = mysql_sequence_id_offset + mysql_sequence_id_size;
+constexpr const auto mysql_payload_offset = mysql_packet_header_size;
+
+bool MySQLHealthChecker::MySQLActiveHealthCheckSession::parse_mysql_server_greeting_packet(
+    Buffer::Instance& data) {
+  // parse server greeting packet:
+  // https://dev.mysql.com/doc/internals/en/connection-phase-packets.html
+  //  #packet-Protocol::Handshake
+
+  // Handshake packet starts with the protocol version in the first byte. Currently we only
+  // support version 10 which has the version number immediately followed by a variable size
+  // NUL terminated string. To minimize linearizations of the buffer we scan for the NUL
+  // terminator.
+  //
+  // TODO (marcelo_juchem): limit the search width to discard malformed packets where the NUL
+  // byte doesn't exist or appears too far into the packet.
+  // `Buffer::search_range(needle, size, start, end)` seems like a useful function to help
+  // avoid performance degradation due to malformed packet attacks.
+  // See `evbuffer_search_range`.
+  const uint8_t nul = '\0';
+  auto server_version_nul = data.search(&nul, 1, mysql_packet_header_size);
+
+  // If NUL not found or at first position in the payload (protocol version 0, which doesn't
+  // exist) the packet is malformed.
+  if (server_version_nul <= static_cast<ssize_t>(mysql_packet_header_size)) {
+    ENVOY_CONN_LOG(trace, "mysql healthcheck failed [greeting]: malformed packet ", *client_);
+    return false;
+  }
+
+  // We calculate the amount of data expected to be in the packet and do a best-effort to
+  // linearize as much as possible in one-shot, but no more.
```

Comment:
What does linearize mean here? Deserialize the payload ? If so I would rephrase accordingly. Because a stream of bytes is inherently linear.

Reply:
linearize is a Buffer operation. I'll check Mongo's implementation as mentioned somewhere else to see how it reads straight from the buffer